src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala

```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
   }
 
   def apply[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)(implicit kOk: OrderedKey[PK, K]): OrderedRDD[PK, K, V] = {
+    val (_, orderedRDD) = coerce(rdd, fastKeys)
+    orderedRDD
+  }
+
+  def coerce[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)
+    (implicit kOk: OrderedKey[PK, K]): (Int, OrderedRDD[PK, K, V]) = {
     import kOk.kct
     import kOk.pkct
     import kOk.kOrd
     import kOk.pkOrd
     import Ordering.Implicits._
 
     if (rdd.partitions.isEmpty)
-      return empty(rdd.sparkContext)
+      return (ORDERED_PARTITIONER, empty(rdd.sparkContext))
 
     rdd match {
-      case ordd: OrderedRDD[PK, K, V] => return ordd
+      case ordd: OrderedRDD[PK, K, V] => return (ORDERED_PARTITIONER, ordd)
       case _ =>
     }
 
     rdd.partitioner match {
-      case Some(op: OrderedPartitioner[PK, K]) => return new OrderedRDD[PK, K, V](rdd, op)
+      case Some(op: OrderedPartitioner[PK, K]) => return (ORDERED_PARTITIONER, new OrderedRDD[PK, K, V](rdd, op))
       case _ =>
     }
 
     val keys = fastKeys.getOrElse(rdd.map(_._1))
 
-    val keyInfoOption = anyFailAllFail[Array, PartitionKeyInfo[PK]](
-      keys.mapPartitionsWithIndex { case (i, it) =>
-        Iterator(if (it.hasNext)
-          Some(PartitionKeyInfo.apply(i, kOk.project, it))
-        else
-          None)
-      }.collect())
-
-    log.info(s"keyInfo = ${ keyInfoOption.map(_.toSeq) }")
-
-    val partitionsSorted =
-      keyInfoOption.exists(keyInfo =>
-        keyInfo.zip(keyInfo.tail).forall { case (p, pnext) =>
-          val r = p.max < pnext.min
-          if (!r)
-            log.info(s"not sorted: p = $p, pnext = $pnext")
-          r
-        })
+    val keyInfo = keys.mapPartitionsWithIndex { case (i, it) =>
+      if (it.hasNext)
+        Iterator(PartitionKeyInfo.apply(i, it))
+      else
+        Iterator()
+    }.collect()
+
+
+    log.info(s"keyInfo = ${ keyInfo.toSeq }")
+
+    if (keyInfo.isEmpty)
+      return (AS_IS, empty(rdd.sparkContext))
+
+    val sortedKeyInfo = keyInfo.sortBy(_.min)
+    val partitionsSorted = sortedKeyInfo.zip(sortedKeyInfo.tail).forall { case (p, pnext) =>
+      val r = p.max < pnext.min
+      if (!r)
+        log.info(s"not sorted: p = $p, pnext = $pnext")
+      r
+    }
 
     if (partitionsSorted) {
-      val keyInfo = keyInfoOption.get
-      val partitioner = OrderedPartitioner(keyInfo.init.map(pi => pi.max))
-      val sortedness = keyInfo.map(_.sortedness).min
+      val partitioner = OrderedPartitioner[PK, K](sortedKeyInfo.init.map(_.max))
+      val sortedness = sortedKeyInfo.map(_.sortedness).min
```

Comment:
This isn't something we can assume.  In particular, if the dataset is locus-sorted but not variant-sorted, we would expect some partitions to be K-sorted and some to be T-sorted.